netplan/cli/commands/generate.py

```diff
@@ -44,6 +46,35 @@ def run(self):
         self.run_command()
 
     def command_generate(self):
+        # if we are inside a snap, then call dbus to run netplan apply instead
+        if "SNAP" in os.environ:
+            # TODO: maybe check if we are inside a classic snap and don't do
+            # this if we are in a classic snap?
+            busctl = shutil.which("busctl")
+            if busctl is None:
+                raise RuntimeError("missing busctl utility")  # pragma: nocover
+            # XXX: DO NOT TOUCH or change this API call, it is used by snapd to communicate
+            #      using core20 netplan binary/client/CLI on core18 base systems. Any change
+            #      must be agreed upon with the snapd team, so we don't break support for
+            #      base systems running older netplan versions.
+            #      https://github.com/snapcore/snapd/pull/10212
+            res = subprocess.call([busctl, "call", "--quiet", "--system",
+                                   "io.netplan.Netplan",  # the service
+                                   "/io/netplan/Netplan",  # the object
+                                   "io.netplan.Netplan",  # the interface
+                                   "Generate",  # the method
+                                   ])
+
+            if res != 0:
+                if res == 130:
+                    raise PermissionError(
+                        "failed to communicate with dbus service")
+                elif res == 1:
+                    raise RuntimeError(
+                        "failed to communicate with dbus service")
+            else:
+                return
```

Comment:
Since I don't see it straight away, but do we have test coverage for this conditional case as well (so res == 0)?

Reply:
Yes. This is the default case, which is being checked in `test_netplan_generate_in_snap_calls_busctl` (https://github.com/canonical/netplan/pull/208/files#diff-2ef3425f71270b61c723558e26aceaa54c5eda80217aa9b216253b146b0754b4R146)

Also `make check-coverage` shows a 100% coverage for C and Python code (the CodeCov service is a bit off here, as it cannot execute all test cases correctly):
```
Overall coverage rate:
  lines......: 100.0% (3639 of 3639 lines)
  functions..: 100.0% (231 of 231 functions)
TOTAL                                     4899      0   100%
```

Reply:
Ah indeed! Was braindead for a moment there. hmm, so since I'm still a bit woozy, what about cases where the error code is != 0 and != 1 and != 130?

Reply:
You're right, that case isn't handled... I was too quick with merging. You can find a followup PR here: https://github.com/canonical/netplan/pull/209